File: app/src/main/java/com/acme/tictactoe/viewmodel/TicTacToeViewModel.java
package com.acme.tictactoe.viewmodel;

import android.databinding.ObservableArrayMap;
import android.databinding.ObservableField;
import android.os.Bundle;

import com.acme.tictactoe.model.Board;
import com.acme.tictactoe.model.Player;

public class TicTacToeViewModel implements ViewModel {

    private static final String SAVED_TTT_VIEWMODEL = "TicTacToeViewModel_SaveBundle";

    private Board model;

    public final ObservableArrayMap<String, String> cells = new ObservableArrayMap<>();
    public final ObservableField<String> winner = new ObservableField<>();

    public TicTacToeViewModel() {
        model = new Board();
    }

    @Override
    public void onCreate() {

    }

    @Override
    public void onPause() {

    }

    @Override
    public void onResume() {

    }

    @Override
    public void onSaveInstanceState(Bundle bundle)
    {
        bundle.putParcelable(SAVED_TTT_VIEWMODEL, model);
    }

    @Override
    public void loadFromSavedInstanceState(Bundle bundle)
    {
        model = bundle.getParcelable(SAVED_TTT_VIEWMODEL);
        // update the view
        if (model != null) {
            for (int row = 0; row < 3; row++) {
                for (int col = 0; col < 3; col++) {
                    if (model.valueAtCell(row, col) != null) {
                        cells.put("" + row + col, model.valueAtCell(row, col).toString());
                    }
                }
            }
            winner.set(model.getWinner() == null ? null : model.getWinner().toString());
        }
    }

    @Override
    public void onDestroy() {

    }

    public void onResetSelected() {
        model.restart();
        winner.set(null);
        cells.clear();
    }

    public void onClickedCellAt(int row, int col) {
        Player playerThatMoved = model.mark(row, col);
        cells.put("" + row + col, playerThatMoved == null ? null : playerThatMoved.toString());
        winner.set(model.getWinner() == null ? null : model.getWinner().toString());
    }

}

File: app/src/main/java/com/acme/tictactoe/viewmodel/TicTacToeViewModel.kt
package com.acme.tictactoe.viewmodel

import android.net.Uri
import androidx.lifecycle.ViewModel
import androidx.lifecycle.liveData
import androidx.lifecycle.viewModelScope
import com.acme.tictactoe.model.Player
import com.acme.tictactoe.model.Symbol
import com.acme.tictactoe.model.TicTacToeGame
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.launch
import java.net.URI

class TicTacToeViewModel (val defaultDispatcher: CoroutineDispatcher = Dispatchers.Main) : ViewModel() {

    private val gameModel = TicTacToeGame(listOf(
            Player("Player 1", Symbol.X),
            Player("Player 2", Symbol.O)))

    @ExperimentalCoroutinesApi
    val gameStateLiveData = liveData(defaultDispatcher) {
        Uri.parse("").fragment
        for(activeGameState in gameModel.gameStatus) {
            emit(activeGameState)
        }
    }

    override fun onCleared() {
        super.onCleared()
        gameModel.shutdown()
    }

    fun onResetSelected() {
        viewModelScope.launch(defaultDispatcher) {
            gameModel.restart()
        }
    }

    fun onClickedCellAt(row: Int, col: Int) {
        viewModelScope.launch(defaultDispatcher) {
            gameModel.mark(row, col)
        }
    }

}
